aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -2646,17 +2646,16 @@ Tensor mexp_impl(
         a_large_norm,
         large_norm_subset,
         thetas[total_n_degs - 1]
-      );
+      ).to(a.dtype());
       res.index_put_({idx_large_norm}, mexp_out);
     }
-
     return res;
   }
 
   return compute_T18_scale_square(
     a, norm,
     thetas[total_n_degs - 1]
-  );
+  ).to(a.dtype());
```

Comment:
I believe we could, but currently I haven't checked it yet.

Reply:
It'd be good if we could do that, yep.

Reply:
Oh, I misunderstood, `compute_T18` will only convert non-contiguous input to contiguous output, instead of changing the dtype.
The `matmul` operation (at line 2576) will change dtype within `torch.x.amp.autocast()` context.

Reply:
Looking at the test, it seems that due to AMP autocasting, the matrix multiplication of fp32 matrices can produce an fp16 result.
This happens in the main branch already when AMP is enabled, but at the end of the matrix_exp computation the fp16 matrix is copied into an fp32 result tensor, so the output appears to be fp32, and the test passes.
When AMP is not enabled, this added `.to(a.dtype())` is a no-op. `compute_T18` already preserves `dtype`.

Reply:
How about (Already tested, it has no significant performance impact):
```python
autocast_state = torch.is_autocast_enabled()
autocast_cpu_state = torch.is_autocast_cpu_enabled()

torch.set_autocast_enabled(False)
torch.set_autocast_cpu_enabled(False)

A = A @ A

torch.set_autocast_enabled(autocast_state)
torch.set_autocast_cpu_enabled(autocast_cpu_state)
```